ruby/lib/google/protobuf/repeated_field.rb

```diff
@@ -29,11 +29,127 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 # add syntatic sugar on top of the core library
+require 'forwardable'
 module Google
   module Protobuf
     class RepeatedField
+      extend Forwardable
 
-      alias_method :size, :length
+      # methods defined in C or Java:
+      #   +
+      #   [], at
+      #   []=
+      #   concat
+      #   clear
+      #   dup, clone
+      #   each
+      #   hash
+      #   push, <<
+      #   replace
+      #   length, size
+      #   ==
+      #   to_ary, to_a
+      #   also all enumerable
+      #
+      # NOTE:  using delegators rather than method_missing to make the
+      #        relationship explicit instead of implicit
+      # NOTE2: use to_a instea
+      def_delegators :to_ary,
+        :&, :*, :-, :'<=>',
+        :assoc, :bsearch, :combination, :compact, :count, :cycle,
+        :drop, :drop_while, :eql?, :fetch, :find_index, :flatten,
+        :include?, :index, :inspect, :join,
+        :pack, :permutation, :product, :pretty_print, :pretty_print_cycle,
+        :rassoc, :repeated_combination, :repeated_permutation, :reverse, :rindex, :rotate,
+        :sample, :shuffle, :shelljoin, :slice, :to_s, :transpose, :uniq, :|
+
+
+      def first(n=nil)
+        n ? self[0..n] : self[0]
+      end
+
+
+      def last(n=nil)
+        n ? self[(self.size-n-1)..-1] : self[-1]
+      end
+
+
+      def pop(n=nil)
+        if n
+          results = []
+          n.times{ results << pop_internal }
+          return results
+        else
+          return pop_internal
+        end
+      end
+
+
+      def empty?
+        self.size == 0
+      end
+
+      # array aliases into enumerable
+      alias_method :each_index, :each_with_index
+      alias_method :slice, :[]
+      alias_method :values_at, :select
+      alias_method :map, :collect
+
+
+      %w(delete delete_at delete_if shift slice! unshift).each do |method_name|
+        define_method(method_name) do |*args, &block|
+          arr = self.to_a
+          result = arr.send(method_name, *args)
+          self.replace(arr)
```

Comment:
yep, I hear ya!  

as I may have mentioned, I tried two things:
* in the c-library, having a ruby array mirror the internal data structure, and use that as a proxy.  That worked fairly well, but I didn't tackle all the edge cases, and it would double, if not more, the memory size...
* I tried implementing these helpers in C *and* java... heh, I'm a glutton for punishment enough as it is without doing that.

I'll put together a benchmark and send it over.  If having parity, or close to parity, is important, than I would say folks can always optimize methods, say, just for C, if they want to, but at least there would be a default implementation and test